platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ResolvedArtifactCollectingVisitor.java

```diff
@@ -45,10 +46,13 @@ public void visitFailure(Throwable failure) {
     @Override
     public void visitArtifact(DisplayName variantName, AttributeContainer variantAttributes, List<? extends Capability> capabilities, ResolvableArtifact artifact) {
         try {
-            if (seenArtifacts.add(artifact.getId())) {
-                File file = artifact.getFile();
-                this.artifacts.add(new DefaultResolvedArtifactResult(artifact.getId(), variantAttributes, capabilities, variantName, Artifact.class, file));
-            }
+            artifacts.compute(new ArtifactIdentifier(artifact.getId(), artifact.getFile()), (key, value) -> {
```

Comment:
I think I agree with sterling in that it seems odd that we're now having one artifact pointing to multiple variants. IMO we should just have two artifacts that point to the same file (and have the same ID) but point to different variants. 

So instead of 
Artifact1 -> [Variant1, Variant2]

It would be
Artifact1 -> Variant1
Artifact1 -> Variant2

In 99.9% of cases an artifact only comes from a single variant and I don't think we should make all those cases more complicated with this approach.

Instead, I'd just go back to the old code, but remove the deduplication (`seenArtifacts.add(...)`) so that if multiple artifacts come in with the same ID we report them.